### PR TITLE
8330153: C2: dump barrier information for all Mach nodes

### PR DESCRIPTION
**Pull request by robcasloz**
This debug-only changeset ensures that GC-specific barrier information is dumped (via [`BarrierSetC2::dump_barrier_data()`](https://github.com/openjdk/jdk/blob/aebfd53e9d19f5939c81fa1a2fc75716c3355900/src/hotspot/share/gc/shared/c2/barrierSetC2.hpp#L311-L313)) for all C2 Mach nodes, not just `MachType` ones. This makes it possible to e.g. write IR tests that verify barrier properties of `CompareAndSwap`/`WeakCompareAndSwap` Mach implementations, which do not inherit from `MachTypeNode`. An example of such a test can be found [here](https://github.com/robcasloz/jdk/blob/e9b3c2a4cb5dd80d85af8320e559acea5920b2ff/test/hotspot/jtreg/compiler/gcbarriers/TestG1BarrierGeneration.java#L283-L292).

The [following program](https://bugs.openjdk.org/secure/attachment/108921/Example.java) illustrates the effect of the change:

```
import java.lang.invoke.VarHandle;
import java.lang.invoke.MethodHandles;

public class Example {
    static class Outer {
        Object f;
    }

    static final VarHandle fVarHandle;
    static {
        MethodHandles.Lookup l = MethodHandles.lookup();
        try {
            fVarHandle = l.findVarHandle(Outer.class, "f", Object.class);
        } catch (Exception e) {
            throw new Error(e);
        }
    }

    static boolean testCompareAndSwap(Outer o, Object oldVal, Object newVal) {
        return fVarHandle.compareAndSet(o, oldVal, newVal);
    }

    public static void main(String[] args) {
        for (int i = 0; i < 10_000; i++) {
            Outer o = new Outer();
            Object oldVal = new Object();
            o.f = oldVal;
            Object newVal = new Object();
            testCompareAndSwap(o, oldVal, newVal);
        }
    }
}
```

Before this changeset, issuing this command:

```
$ java -Xbatch -XX:+UseZGC -XX:+ZGenerational -XX:CompileOnly=Example::testCompareAndSwap -XX:CompileCommand=PrintIdealPhase,Example::testCompareAndSwap,FINAL_CODE Example.java | grep zCompareAndSwapP
```

gives the following dump:

```
 R10     37  zCompareAndSwapP  === 28 35 38 54 22 41  [[ 40 42 36 27 56 25 ]]  !jvms: VarHandleReferences$FieldInstanceReadWrite::compareAndSet @ bci:44 (line 180) VarHandleGuards::guard_LLL_Z @ bci:50 (line 68) Example::testCompareAndSwap @ bci:6 (line 20)
```

After this changeset, we get:

```
 R10     37  zCompareAndSwapP  === 28 35 38 54 22 41  [[ 40 42 36 27 56 25 ]]  barrier(strong )  !jvms: VarHandleReferences$FieldInstanceReadWrite::compareAndSet @ bci:44 (line 180) VarHandleGuards::guard_LLL_Z @ bci:50 (line 68) Example::testCompareAndSwap @ bci:6 (line 20)
```

Note the additional `barrier(strong )` field in the second dump.

**Testing:** tier1-3 (windows-x64, linux-x64, linux-aarch64, and macosx-x64; release and debug mode).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330153](https://bugs.openjdk.org/browse/JDK-8330153): C2: dump barrier information for all Mach nodes (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [409a1ef4](https://git.openjdk.org/jdk/pull/18754/files/409a1ef4cc86b3cd591ce84c7f5e305115d3a101)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18754/head:pull/18754` \
`$ git checkout pull/18754`

Update a local copy of the PR: \
`$ git checkout pull/18754` \
`$ git pull https://git.openjdk.org/jdk.git pull/18754/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18754`

View PR using the GUI difftool: \
`$ git pr show -t 18754`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18754.diff">https://git.openjdk.org/jdk/pull/18754.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18754#issuecomment-2051746786)

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Can you add a simple IR test to show case it?

**Comment by robcasloz** [comment]
> Can you add a simple IR test to show case it?

Done, the new test fails before this change and passes afterwards.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by robcasloz** [comment]
Thanks for reviewing, Tobias!

@vnkozlov: I added a simple IR test as requested, could I get a second review? Thanks!

**Comment by vnkozlov** [comment]
I don't see GHA testing for this repo. Did you enabled it?

File: test/hotspot/jtreg/testlibrary_tests/ir_framework/examples/GCBarrierIRExample.java
Comment (vnkozlov) [inline test/hotspot/jtreg/testlibrary_tests/ir_framework/examples/GCBarrierIRExample.java:31]: Missing `@bug`

Comment (robcasloz) [reply test/hotspot/jtreg/testlibrary_tests/ir_framework/examples/GCBarrierIRExample.java:31]: Fixed, thanks.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by robcasloz** [comment]
> I don't see GHA testing for this repo. Did you enabled it?

Interesting. GHA testing is enabled for the repo (see https://github.com/robcasloz/jdk/actions), but it seems to run only for some branches, see e.g. my recent PR #18108 from the same repo. I will investigate.

**Comment by robcasloz** [comment]
Thanks for reviewing, Vladimir!

**Pull request closed without merging**
